<?xml version='1.0' encoding='UTF-8'?>
<root>
 <tabbed_widget name="Main Window" parent="main_window">
  <Tab tab_name="tab1" containers="1">
   <Container>
    <DockSplitter count="3" orientation="-" sizes="0.333333;0.333333;0.333333">
     <DockArea name="...">
      <plot flip_y="false" style="Lines" flip_x="false" mode="TimeSeries">
       <range left="6.137000" top="1.680404" bottom="-2.921306" right="51.137000"/>
       <limitY/>
       <curve name="/joint_states/ur5_elbow_joint/position" color="#1f77b4"/>
       <curve name="/joint_states/ur5_shoulder_lift_joint/position" color="#f14cc1"/>
       <curve name="/joint_states/ur5_shoulder_pan_joint/position" color="#bcbd22"/>
       <curve name="/joint_states/ur5_wrist_1_joint/position" color="#1ac938"/>
       <curve name="/joint_states/ur5_wrist_2_joint/position" color="#9467bd"/>
       <curve name="/joint_states/ur5_wrist_3_joint/position" color="#1f77b4"/>
      </plot>
     </DockArea>
     <DockArea name="...">
      <plot flip_y="false" style="Lines" flip_x="false" mode="TimeSeries">
       <range left="6.137000" top="0.000001" bottom="-0.000001" right="51.137000"/>
       <limitY/>
       <curve name="/joint_states/ur5_elbow_joint/velocity" color="#d62728"/>
       <curve name="/joint_states/ur5_shoulder_lift_joint/velocity" color="#9467bd"/>
       <curve name="/joint_states/ur5_shoulder_pan_joint/velocity" color="#1f77b4"/>
       <curve name="/joint_states/ur5_wrist_1_joint/velocity" color="#ff7f0e"/>
       <curve name="/joint_states/ur5_wrist_2_joint/velocity" color="#17becf"/>
       <curve name="/joint_states/ur5_wrist_3_joint/velocity" color="#d62728"/>
      </plot>
     </DockArea>
     <DockArea name="...">
      <plot flip_y="false" style="Lines" flip_x="false" mode="TimeSeries">
       <range left="6.137000" top="83.382370" bottom="-2.110467" right="51.137000"/>
       <limitY/>
       <curve name="/joint_states/ur5_elbow_joint/effort" color="#1ac938"/>
       <curve name="/joint_states/ur5_shoulder_lift_joint/effort" color="#ff7f0e"/>
       <curve name="/joint_states/ur5_shoulder_pan_joint/effort" color="#17becf"/>
       <curve name="/joint_states/ur5_wrist_1_joint/effort" color="#d62728"/>
       <curve name="/joint_states/ur5_wrist_2_joint/effort" color="#f14cc1"/>
       <curve name="/joint_states/ur5_wrist_3_joint/effort" color="#bcbd22"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <currentTabIndex index="0"/>
 </tabbed_widget>
 <use_relative_time_offset enabled="1"/>
 <!-- - - - - - - - - - - - - - - -->
 <!-- - - - - - - - - - - - - - - -->
 <Plugins>
  <plugin ID="DataLoad CSV">
   <default time_axis="" delimiter="0"/>
  </plugin>
  <plugin ID="DataLoad MCAP"/>
  <plugin ID="DataLoad ROS bags">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
  </plugin>
  <plugin ID="DataLoad ULog"/>
  <plugin ID="ROS Topic Subscriber">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="false"/>
   <remove_suffix_from_strings value="false"/>
  </plugin>
  <plugin ID="UDP Server"/>
  <plugin ID="WebSocket Server"/>
  <plugin ID="ZMQ Subscriber"/>
  <plugin ID="Fast Fourier Transform"/>
  <plugin ID="Quaternion to RPY"/>
  <plugin ID="Reactive Script Editor">
   <library code="--[[ Helper function to create a series from arrays&#xa;&#xa; new_series: a series previously created with ScatterXY.new(name)&#xa; prefix:     prefix of the timeseries, before the index of the array&#xa; suffix_X:   suffix to complete the name of the series containing the X value. If [nil], use the index of the array.&#xa; suffix_Y:   suffix to complete the name of the series containing the Y value&#xa; timestamp:   usually the tracker_time variable&#xa;              &#xa; Example:&#xa; &#xa; Assuming we have multiple series in the form:&#xa; &#xa;   /trajectory/node.{X}/position/x&#xa;   /trajectory/node.{X}/position/y&#xa;   &#xa; where {N} is the index of the array (integer). We can create a reactive series from the array with:&#xa; &#xa;   new_series = ScatterXY.new(&quot;my_trajectory&quot;) &#xa;   CreateSeriesFromArray( new_series, &quot;/trajectory/node&quot;, &quot;position/x&quot;, &quot;position/y&quot;, tracker_time );&#xa;--]]&#xa;&#xa;function CreateSeriesFromArray( new_series, prefix, suffix_X, suffix_Y, timestamp )&#xa;  &#xa;  --- clear previous values&#xa;  new_series:clear()&#xa;  &#xa;  --- Append points to new_series&#xa;  index = 0&#xa;  while(true) do&#xa;&#xa;    x = index;&#xa;    -- if not nil, get the X coordinate from a series&#xa;    if suffix_X ~= nil then &#xa;      series_x = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_X) )&#xa;      if series_x == nil then break end&#xa;      x = series_x:atTime(timestamp)&#x9; &#xa;    end&#xa;    &#xa;    series_y = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_Y) )&#xa;    if series_y == nil then break end &#xa;    y = series_y:atTime(timestamp)&#xa;    &#xa;    new_series:push_back(x,y)&#xa;    index = index+1&#xa;  end&#xa;end&#xa;&#xa;--[[ Similar to the built-in function GetSeriesNames(), but select only the names with a give prefix. --]]&#xa;&#xa;function GetSeriesNamesByPrefix(prefix)&#xa;  -- GetSeriesNames(9 is a built-in function&#xa;  all_names = GetSeriesNames()&#xa;  filtered_names = {}&#xa;  for i, name in ipairs(all_names)  do&#xa;    -- check the prefix&#xa;    if name:find(prefix, 1, #prefix) then&#xa;      table.insert(filtered_names, name);&#xa;    end&#xa;  end&#xa;  return filtered_names&#xa;end&#xa;&#xa;--[[ Modify an existing series, applying offsets to all their X and Y values&#xa;&#xa; series: an existing timeseries, obtained with TimeseriesView.find(name)&#xa; delta_x: offset to apply to each x value&#xa; delta_y: offset to apply to each y value &#xa;  &#xa;--]]&#xa;&#xa;function ApplyOffsetInPlace(series, delta_x, delta_y)&#xa;  -- use C++ indeces, not Lua indeces&#xa;  for index=0, series:size()-1 do&#xa;    x,y = series:at(index)&#xa;    series:set(index, x + delta_x, y + delta_y)&#xa;  end&#xa;end&#xa;"/>
   <scripts/>
  </plugin>
  <plugin ID="CSV Exporter"/>
  <plugin ID="ROS /rosout Visualization"/>
  <plugin ID="ROS Topic Re-Publisher"/>
 </Plugins>
 <!-- - - - - - - - - - - - - - - -->
 <previouslyLoaded_Datafiles/>
 <previouslyLoaded_Streamer name="ROS Topic Subscriber"/>
 <!-- - - - - - - - - - - - - - - -->
 <customMathEquations/>
 <snippets>
  <snippet name="1st_derivative">
   <global>var prevX = 0
var prevY = 0</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="1st_order_lowpass">
   <global>var prevY = 0
var alpha = 0.1</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="sum_A_B">
   <global></global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="yaw_from_quaternion">
   <global>// source: https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles

function quaternionToYaw(x, y, z, w)
{
  // yaw (z-axis rotation)
  t1 = 2.0 * (w * z + x * y);
  t2 = 1.0 - 2.0 * (y * y + z * z);
  yaw = Math.atan2(t1, t2);

  return yaw
}</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
 </snippets>
 <!-- - - - - - - - - - - - - - - -->
</root>

